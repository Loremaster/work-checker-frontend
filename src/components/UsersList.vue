<template>
  <v-data-table :headers="headers" :items="users" class="elevation-1">
    <template v-slot:items="props">
      <td>{{ props.item.email }}</td>
      <td>{{ props.item.role }}</td>
    </template>
  </v-data-table>
</template>

<script lang="ts">
import { Component, Vue, Watch } from "vue-property-decorator";
import { Action, Getter } from "vuex-class";
import { IUserList } from "@/interfaces";

@Component
export default class UsersList extends Vue {
  headers = [{ text: "Email", value: "email" }, { text: "Role", value: "role" }];

  @Action("fetchUsers") fetchUsers: any;

  @Getter("users") users!: IUserList;

  async created() {
    await this.fetchUsers();
  }
}
</script>
