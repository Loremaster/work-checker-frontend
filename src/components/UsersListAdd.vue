<template>
  <v-form>
    <v-container>
      <v-layout>
        <v-flex md4>
          <v-text-field
            v-model="email"
            v-validate="'required|email'"
            :error-messages="errors.collect('email')"
            name="email"
            label="Email"
            data-vv-name="email"
            required
          />
        </v-flex>
        <v-flex md3>
          <v-text-field
            v-model="password"
            v-validate="'required|min:10'"
            :error-messages="errors.collect('password')"
            name="password"
            label="Password"
            data-vv-name="password"
            required
          />
        </v-flex>
        <v-flex md3>
          <v-select
            v-model="role"
            v-validate="'required'"
            :items="roles"
            :error-messages="errors.collect('role')"
            name="role"
            label="Role"
            data-vv-name="role"
            required
          />
        </v-flex>
        <v-flex md1>
          <v-btn @click="$emit('cancel')">Cancel</v-btn>
        </v-flex>
        <v-flex md1>
          <v-btn ref="createBtn" color="primary" @click="submit">Create</v-btn>
        </v-flex>
      </v-layout>
    </v-container>
  </v-form>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import { Action } from "vuex-class";
import VeeValidate from "vee-validate";

Vue.use(VeeValidate);

@Component({
  $_veeValidate: { validator: "new" },
})
export default class UsersListAdd extends Vue {
  email: string = "";
  role: string = "";
  password: string = "";
  roles: string[] = ["farmer", "admin"];

  @Action("createUser") createUser: any;

  reset() {
    this.email = "";
    this.password = "";
    this.role = "";
  }

  async submit() {
    const isValid: boolean = await this.$validator.validateAll();

    if (isValid) {
      await this.crateUser();
    }
  }

  async crateUser() {
    try {
      await this.createUser({ email: this.email, password: this.password, role: this.role });
      this.reset();
      this.$emit("created");
    } catch (e) {
      const errorsData: { [k: string]: string[] } = e.response.data;

      for (const [field, messages] of Object.entries(errorsData)) {
        const message: string = messages.join(", ");
        this.errors.add({ field, msg: message });
      }
    }
  }
}
</script>
