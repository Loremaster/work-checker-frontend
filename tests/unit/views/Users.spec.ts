import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import Vuetify from "vuetify";
import UsersList from "@/components/UsersList.vue";
import UsersListAdd from "@/components/UsersListAdd.vue";
import Component from "@/views/Users.vue";

Vue.use(Vuetify);

describe("Authentication.vue", () => {
  const setup = () => {
    return shallowMount(Component);
  };

  it("renders correctly", () => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("renders UsersList component", () => {
    const wrapper = setup();
    const signin = wrapper.find(UsersList);

    expect(signin.is(UsersList)).toEqual(true);
  });

  describe("UsersListAdd", () => {
    it("renders component on click add button", () => {
      const wrapper = setup();

      wrapper.find({ ref: "showForm" }).vm.$emit("click");
      const usersListAdd = wrapper.find(UsersListAdd);

      expect(usersListAdd.is(UsersListAdd)).toEqual(true);
    });

    it("hides component on created event", () => {
      const wrapper = setup();
      wrapper.setData({ creating: true });
      const usersListAdd = wrapper.find(UsersListAdd);

      usersListAdd.vm.$emit("created");

      expect(wrapper.find(UsersListAdd).exists()).toEqual(false);
    });

    it("hides component on cancel event", () => {
      const wrapper = setup();
      wrapper.setData({ creating: true });
      const usersListAdd = wrapper.find(UsersListAdd);

      usersListAdd.vm.$emit("cancel");

      expect(wrapper.find(UsersListAdd).exists()).toEqual(false);
    });
  });
});
