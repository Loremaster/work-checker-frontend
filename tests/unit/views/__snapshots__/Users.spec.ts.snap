// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication.vue renders correctly 1`] = `
<v-container-stub tag="div" grid-list-md="">
  <v-layout-stub tag="div" row="" wrap="">
    <v-flex-stub tag="div" md2="" offset-md10="" class="text-xs-right">
      <v-btn-stub color="primary" activeclass="v-btn--active" tag="button" type="button" class="mr-0">
        Add user
      </v-btn-stub>
    </v-flex-stub>
    <!---->
    <v-flex-stub tag="div" md12="">
      <users-list-stub></users-list-stub>
    </v-flex-stub>
  </v-layout-stub>
</v-container-stub>
`;
